src/audio/sr_audio/sr_audio.c

```diff
@@ -0,0 +1,720 @@
+  struct comp_data* cd = comp_get_drvdata(dev);
+  comp_info(dev, "sraudio_free()");
+  comp_data_blob_handler_free(cd->model_handler);
+  rfree(cd);
+  rfree(dev);
+}
+
+static int sraudio_verify_params(struct comp_dev* dev,
+  struct sof_ipc_stream_params* params) {
+  struct comp_buffer* sourceb;
+  struct comp_buffer* sinkb;
+  uint32_t buffer_flag;
+  int ret;
+  comp_dbg(dev, "sraudio_verify_params()");
+  sourceb = list_first_item(&dev->bsource_list, struct comp_buffer, sink_list);
+  sinkb = list_first_item(&dev->bsink_list, struct comp_buffer, source_list);
+  buffer_flag = sraudio_find_func(sourceb->stream.frame_fmt, sinkb->stream.frame_fmt, fm_configured, ARRAY_SIZE(fm_configured)) ? BUFF_PARAMS_FRAME_FMT : 0;
+  ret = comp_verify_params(dev, buffer_flag, params);
+  if (ret < 0) {
+    comp_err(dev, "sraudio_verify_params(): comp_verify_params() failed.");
+    return ret;
+  }
+  return 0;
+}
+
+static int sraudio_params(struct comp_dev* dev, struct sof_ipc_stream_params* params) {
+  int err;
+  comp_info(dev, "sraudio_params()");
+  err = sraudio_verify_params(dev, params);
+  if (err < 0) {
+    comp_err(dev, "sraudio_params(): pcm params verification failed.");
+    return -EINVAL;
+  }
+  return 0;
+}
+
+static int sraudio_cmd_get_data(struct comp_dev* dev, struct sof_ipc_ctrl_data* cdata, int max_size) {
+  struct comp_data* cd = comp_get_drvdata(dev);
+  int ret = 0;
+  switch (cdata->cmd) {
+  case SOF_CTRL_CMD_BINARY:
+    comp_info(dev, "sraudio_cmd_get_data(), SOF_CTRL_CMD_BINARY");
+    ret = comp_data_blob_get_cmd(cd->model_handler, cdata,
+      max_size);
+    break;
+  default:
+    comp_err(dev, "sraudio_cmd_get_data(), invalid command");
+    ret = -EINVAL;
+    break;
+  }
+  return ret;
+}
+
+static int sraudio_cmd_set_data(struct comp_dev* dev, struct sof_ipc_ctrl_data* cdata) {
+  struct comp_data* cd = comp_get_drvdata(dev);
+  int ret = 0;
+  switch (cdata->cmd) {
+  case SOF_CTRL_CMD_BINARY:
+    comp_info(dev, "sraudio_cmd_set_data(), SOF_CTRL_CMD_BINARY");
+    ret = comp_data_blob_set_cmd(cd->model_handler, cdata);
+    break;
+  default:
+    comp_err(dev, "sraudio_cmd_set_data(), invalid command");
+    ret = -EINVAL;
+    break;
+  }
+  return ret;
+}
+
+static int sraudio_cmd(struct comp_dev* dev, int cmd, void* data, int max_data_size) {
+  struct sof_ipc_ctrl_data* cdata = ASSUME_ALIGNED(data, 4);
+  int ret = 0;
+  comp_info(dev, "sraudio_cmd()");
+  switch (cmd) {
+  case COMP_CMD_SET_DATA:
+    ret = sraudio_cmd_set_data(dev, cdata);
+    break;
+  case COMP_CMD_GET_DATA:
+    ret = sraudio_cmd_get_data(dev, cdata, max_data_size);
+    break;
+  default:
+    comp_err(dev, "sraudio_cmd(), invalid command");
+    ret = -EINVAL;
+  }
+  return ret;
+}
+
+static int sraudio_trigger(struct comp_dev* dev, int cmd) {
+  struct comp_data* cd = comp_get_drvdata(dev);
+  comp_info(dev, "sraudio_trigger()");
+  if (cmd == COMP_TRIGGER_START || cmd == COMP_TRIGGER_RELEASE)
+    assert(cd->sraudio_func);
```

Comment:
you should be doing this check when you prepare or your params step when you actually look up this func, also don't assert, return an error